src/hardware/mixer.cpp

```diff
@@ -576,9 +576,11 @@ spx_uint32_t estimate_max_out_frames(SpeexResamplerState *resampler_state,
 template <class Type, bool stereo, bool signeddata, bool nativeorder>
 void MixerChannel::AddSamples(const uint16_t frames, const Type *data)
 {
-	assert(frames > 0);
 	assert(frames <= MIXER_BUFSIZE);
 
+	if (!frames)
+		return;
+
```

Comment:
Can you tell which audio device is still calling `AddSamples(..)` with zero?
If so, let's try to fix it in there instead of here.

Reply:
I'm not too sure about asserting the max buffer size either. 

Indeed. `assert(frames <= MIXER_BUFSIZE);`, can go because we're not using fixed-size arrays anymore. 

The vector's will accommodate beyond `MIXER_BUFSIZE` if that's actually the case. 

16384 frames / 48 frames/ms = 341 ms of deficit, which is definitely possible on machines that get into trouble.. like a huge IO or swap storm.. I've seen Linux starve processes for a lot longer than a half-millisecond! So this is a good thing (moving to vectors). Granted.. if DOSbox is hung for half a second there are going to be major audio and visual fall out anyway, but not much we can do.

> I'm not 100% sure if these asserts are disabled in release builds. If yes, it's less of a problem

They're positively knocked out in release builds; @dreamer and I made sure of that with hard verifications (add `assert(false);` to confirm it). We needed that assurance before using them as assumption-checks throughout the newly added code.

In this case, the code is still "zero-tolerant"; the calculation, `0 / something`, returns a max-size of zero. Both the in-bound and out-bound frames counts passed into the resampler are zero, so it does nothing, as does the rest of the function.

> It's just too risky to leave that assert there. The end user doesn't care, they just don't want their game to crash!

The good news is the release-builds have already been traversing this zero pathway before we added the assert. It's our expectation though (and verified in debug mode) that `AddSamples` is never passed zero samples -- so it's a good thing we're finally rooting out these cases.

The loop that call `AddSamples()` should guarantee it, so it will be easy to track down the remaining culprits and fix them.

Reply:
The main branch is meant to help us as maintainers - and users of it know that. So if there's any change of the non-assert pathway failing, that's what this is for.

Suggest we leave the assert as is, and get a back trace to find which audio device is calling it.

If we're still hitting these assert come 0.79 release time, worst case we can add if the `if & return` as a safety net.

Reply:
My desire here isn't able some micro-optimization or the extra lines of code (if we need a check, then we need it).

But it's more about discovering that we have two parts of code in a battle - entirely under out control (not user/data side), where one side breaks and the other side fixes it up.  When it's discovered, I want to keep sorting that out until it's solved (and have the assert in there to "prove it" going forward).

The AddSamples scenarios are quite understandable, so I just want to get to the bottom of it.

Reply:
Thanks for the clarification. I don't think we're disagreeing, I just wasn't quite sure about the role of the asserts.

> The vectors will accommodate beyond MIXER_BUFSIZE if that's actually the case.

Yeah, good point, removed.

> The good news is the release builds have already been traversing this zero pathway before we added the assert.

Yeah, that's the key thing for me, and I kinda forgot about it. I fully agree that the asserts are very useful during dev to catch problems early on (I do that myself a lot in other projects of mine, but only at debug time). I just wasn't sure they're disabled in the real release builds, but if that's the case, my concerns have been addressed 😄 

I guess what I'm advocating is doing the asserts first, and _also_ do things defensively so when the asserts are disabled in the release build things won't just blow up (you might not hit the asserts during your dev testing, but some user somewhere might, and won't be happy if the game crashes). But that's more of a general comment, `AddSamples` seems to be handling the zero frame situation fine already. ✅ 

With the rest, I agree — sure, it's better to fix problems at the root. All I'm saying in many cases it's ultra hard to anticipate how things are exactly called in every single situation, so better to be defensive as well.

> The main branch is meant to help us as maintainers - and users of it know that. So if the non-assert (release) pathway fails, then that's what main is there for - and users will tell us.

I'm slightly confused here, but then I don't fully understand the intricacies of the build system (yet), so I'm actually not too surprised 😂 On Windows I'm just doing "Release" builds in VC2019, and on OS X I just followed the build instructions from the readme (I'm simply building with "ninja -C build"). So seems like on Windows I'm getting a real release build, and on OS X... the debug build, I guess? Also, I thought there's nothing special about releases; we just slap a tag on some commit in main and that's the release... or there's something else going on? So yeah, slightly confused.

Reply:
> Can you tell which audio device is still calling AddSamples(..) with zero?

It was the SBPro1, and I tried reproducing it now like 20 times in a row, to no avail 🤷🏻 But the crash is still in my scrollback buffer, pasting the full log here. This was while experimenting with my next crossfeed patch; as you can see I've done the "start / exit / adjust config params / start ..." loop a few times. The crash happened immediately after pressing F1 to load the Ignition table (until then I just exited a few seconds after the intro music started playing). I was using the default config, no customisations whatsoever (apart from tweaking the crossfeed param between game runs, but that has no implications on the number of frames passed into `AddSamples`).

```
% ~/dev/jn/dosbox-staging/build/dosbox                                                            [cmd]
date       time         |
2022-05-28 11:16:41.439 | arguments: /Users/jnovak/dev/jn/dosbox-staging/build/dosbox
2022-05-28 11:16:41.439 | Current dir: /Users/jnovak/Documents/DOS/Pinball Dreams
2022-05-28 11:16:41.439 | stderr verbosity: 0
2022-05-28 11:16:41.439 | -----------------------------------
2022-05-28 11:16:41.441 | dosbox-staging version 0.79.0-alpha-501-g80b0b0ca+
2022-05-28 11:16:41.441 | ---
2022-05-28 11:16:41.441 | LOG: Loguru version 2.1.0 initialized
2022-05-28 11:16:41.531 | SDL: version 2.0.22 initialized (cocoa video and coreaudio audio)
2022-05-28 11:16:41.535 | CONFIG: Loaded primary conf file /Users/jnovak/Library/Preferences/DOSBox/dosbox-staging.conf
2022-05-28 11:16:41.535 | CONFIG: Loaded local conf file dosbox.conf
2022-05-28 11:16:41.536 | RENDER: Using GLSL shader 'sharp'
2022-05-28 11:16:41.540 | DISPLAY: Initialized 1106x829 window-mode using Bilinear scaling on 1120p display-0
2022-05-28 11:16:41.586 | OPENGL: Vendor: ATI Technologies Inc.
2022-05-28 11:16:41.586 | OPENGL: Version: 2.1 ATI-4.6.21
2022-05-28 11:16:41.586 | OPENGL: GLSL version: 1.20
2022-05-28 11:16:41.586 | OPENGL: Pixel buffer object: available
2022-05-28 11:16:41.586 | OPENGL: NPOT textures: supported
2022-05-28 11:16:41.587 | SDL: Using standard SDI (auto) display refresh rate of 60 Hz
2022-05-28 11:16:41.587 | DISPLAY: Unknown 640x400 (FFFFh) at 70 Hz synced CFR, scaled by 1.7x1.7 to 1106x691 with 1.0 pixel-aspect
2022-05-28 11:16:42.624 | SDL: Mouse will move seamlessly without being captured and middle-click will uncapture the mouse.
2022-05-28 11:16:42.624 | LANG: Using internal English language messages
2022-05-28 11:16:42.624 | VIDEO: Using the DOS video mode's frame rate
2022-05-28 11:16:42.624 | SYSTEM: Speed modifications are enabled
2022-05-28 11:16:42.649 | MEMORY: Base address: 0x7f8cb2500000
2022-05-28 11:16:42.649 | MEMORY: Using 4096 DOS memory pages (16 MiB)
2022-05-28 11:16:42.651 | VIDEO: Initialized S3 Trio64 VESA 2.0 compatible with 4-MiB of FP DRAM supporting 86 modes
2022-05-28 11:16:42.744 | MIXER: Negotiated 2-channel 48000-Hz audio in 512-frame blocks
2022-05-28 11:16:42.752 | MIDI: Opened device: coreaudio
2022-05-28 11:16:42.753 | MIXER: CMS channel operating at 48000 Hz without resampling
2022-05-28 11:16:42.755 | CMS: Running on port 220h with two 7.159 MHz Phillips SAA-1099 chips
2022-05-28 11:16:42.755 | MIXER: SB channel operating at 22050 Hz and upsampling to the output rate
2022-05-28 11:16:42.755 | SBPRO1: @SET BLASTER=A220 I7 D1 T2
2022-05-28 11:16:42.755 | MIXER: SPKR channel operating at 18939 Hz and upsampling to the output rate
2022-05-28 11:16:42.755 | MIXER: DISNEY channel operating at 10000 Hz and upsampling to the output rate
2022-05-28 11:16:42.784 | SLIRP: Slirp version: 4.6.1
2022-05-28 11:16:42.784 | SLIRP: Successfully initialized
2022-05-28 11:16:42.784 | NE2000: Base=0x300 irq=3
2022-05-28 11:16:42.784 | AUTOEXEC: Using autoexec from one or more joined sections
2022-05-28 11:16:42.800 | MAPPER: no joysticks found
2022-05-28 11:16:42.800 | MAPPER: Can't find key binding for key_rctrl event
2022-05-28 11:16:42.800 | MAPPER: Loaded default key bindings
2022-05-28 11:16:42.807 | MOUNT: Path 'C' found
2022-05-28 11:16:42.822 | DISPLAY: Text 640x400 (3h) at 70.087 Hz VFR, scaled by 1.7x2.1 to 1105x829 with 1.2 pixel-aspect
2022-05-28 11:16:43.145 | SDL: Window has been shown
2022-05-28 11:16:49.854 | SetCrossfeedAmount: 0.500000
2022-05-28 11:16:49.854 | SetCrossfeedAmount pan_left: 0.250000
2022-05-28 11:16:49.854 | SetCrossfeedAmount pan_right: 0.750000
2022-05-28 11:16:49.881 | PACER: Render took 23383us, skipping next
2022-05-28 11:16:50.834 | DISPLAY: VGA 320x200 8-bit (13h) at 70.086 Hz VFR, scaled by 3.5x4.1 to 1105x829 with 1.2 pixel-aspect
2022-05-28 11:16:54.251 | SBPRO1: DSP was reset
2022-05-28 11:16:54.251 | SBPRO1: Speaker-output has been toggled on
2022-05-28 11:16:57.031 | ENVELOPE: SB done after 121951 frames, peak sample was 29700
2022-05-28 11:17:13.290 | SBPRO1: DSP was reset
2022-05-28 11:17:13.290 | SBPRO1: Speaker-output has been toggled on
2022-05-28 11:17:13.291 | SBPRO1: Speaker-output has been toggled off
2022-05-28 11:17:13.871 | DISPLAY: Text 640x400 (3h) at 70.087 Hz VFR, scaled by 1.7x2.1 to 1105x829 with 1.2 pixel-aspect
2022-05-28 11:17:17.142 | SetCrossfeedAmount: 0.700000
2022-05-28 11:17:17.142 | SetCrossfeedAmount pan_left: 0.350000
2022-05-28 11:17:17.142 | SetCrossfeedAmount pan_right: 0.650000
2022-05-28 11:17:18.404 | DISPLAY: VGA 320x200 8-bit (13h) at 70.086 Hz VFR, scaled by 3.5x4.1 to 1105x829 with 1.2 pixel-aspect
2022-05-28 11:17:21.802 | SBPRO1: DSP was reset
2022-05-28 11:17:21.803 | SBPRO1: Speaker-output has been toggled on
2022-05-28 11:17:24.463 | SBPRO1: DSP was reset
2022-05-28 11:17:24.463 | SBPRO1: Speaker-output has been toggled on
2022-05-28 11:17:24.463 | SBPRO1: Speaker-output has been toggled off
2022-05-28 11:17:25.043 | DISPLAY: Text 640x400 (3h) at 70.087 Hz VFR, scaled by 1.7x2.1 to 1105x829 with 1.2 pixel-aspect
2022-05-28 11:17:27.264 | SetCrossfeedAmount: 1.000000
2022-05-28 11:17:27.264 | SetCrossfeedAmount pan_left: 0.500000
2022-05-28 11:17:27.264 | SetCrossfeedAmount pan_right: 0.500000
2022-05-28 11:17:27.858 | DISPLAY: VGA 320x200 8-bit (13h) at 70.086 Hz VFR, scaled by 3.5x4.1 to 1105x829 with 1.2 pixel-aspect
2022-05-28 11:17:31.263 | SBPRO1: DSP was reset
2022-05-28 11:17:31.263 | SBPRO1: Speaker-output has been toggled on
2022-05-28 11:17:34.922 | SBPRO1: DSP was reset
2022-05-28 11:17:34.922 | SBPRO1: Speaker-output has been toggled on
2022-05-28 11:17:34.922 | SBPRO1: Speaker-output has been toggled off
2022-05-28 11:17:35.502 | DISPLAY: Text 640x400 (3h) at 70.087 Hz VFR, scaled by 1.7x2.1 to 1105x829 with 1.2 pixel-aspect
2022-05-28 11:17:42.055 | SetCrossfeedAmount: 0.300000
2022-05-28 11:17:42.055 | SetCrossfeedAmount pan_left: 0.150000
2022-05-28 11:17:42.055 | SetCrossfeedAmount pan_right: 0.850000
2022-05-28 11:17:45.860 | DISPLAY: VGA 320x200 8-bit (13h) at 70.086 Hz VFR, scaled by 3.5x4.1 to 1105x829 with 1.2 pixel-aspect
2022-05-28 11:17:49.426 | SBPRO1: DSP was reset
2022-05-28 11:17:49.426 | SBPRO1: Speaker-output has been toggled on
2022-05-28 11:17:52.171 | SBPRO1: DSP was reset
2022-05-28 11:17:52.171 | SBPRO1: Speaker-output has been toggled on
2022-05-28 11:17:52.171 | SBPRO1: Speaker-output has been toggled off
2022-05-28 11:17:52.752 | DISPLAY: Text 640x400 (3h) at 70.087 Hz VFR, scaled by 1.7x2.1 to 1105x829 with 1.2 pixel-aspect
2022-05-28 11:17:55.077 | SetCrossfeedAmount: 0.000000
2022-05-28 11:17:55.077 | SetCrossfeedAmount pan_left: 0.000000
2022-05-28 11:17:55.077 | SetCrossfeedAmount pan_right: 1.000000
2022-05-28 11:17:56.210 | DISPLAY: VGA 320x200 8-bit (13h) at 70.086 Hz VFR, scaled by 3.5x4.1 to 1105x829 with 1.2 pixel-aspect
2022-05-28 11:17:59.601 | SBPRO1: DSP was reset
2022-05-28 11:17:59.601 | SBPRO1: Speaker-output has been toggled on
2022-05-28 11:18:01.646 | SBPRO1: DSP was reset
2022-05-28 11:18:01.646 | SBPRO1: Speaker-output has been toggled on
2022-05-28 11:18:01.646 | SBPRO1: Speaker-output has been toggled off
2022-05-28 11:18:02.227 | DISPLAY: Text 640x400 (3h) at 70.087 Hz VFR, scaled by 1.7x2.1 to 1105x829 with 1.2 pixel-aspect
2022-05-28 11:18:04.364 | SetCrossfeedAmount: 0.400000
2022-05-28 11:18:04.364 | SetCrossfeedAmount pan_left: 0.200000
2022-05-28 11:18:04.364 | SetCrossfeedAmount pan_right: 0.800000
2022-05-28 11:18:05.011 | DISPLAY: VGA 320x200 8-bit (13h) at 70.086 Hz VFR, scaled by 3.5x4.1 to 1105x829 with 1.2 pixel-aspect
2022-05-28 11:18:08.318 | SBPRO1: DSP was reset
2022-05-28 11:18:08.318 | SBPRO1: Speaker-output has been toggled on
Assertion failed: (frames > 0), function AddSamples, file mixer.cpp, line 613.

Loguru caught a signal: SIGABRT
2022-05-28 11:18:52.457 | Signal: SIGABRT
zsh: abort      ~/dev/jn/dosbox-staging/build/dosbox
```

Reply:
Setting to resolved; we have a way forward on this (as currently represented in the code), and the issue is also out-of-scope from the PR.